File: scala/pal/src/main/scala/pal/ScaffoldGsFlags.scala
package pal

import scala.collection.immutable.VectorMap
import scala.collection.mutable
import Expr.{TRUE, FALSE}
import Ref.{NEW, PREVIOUS}
import ScaffoldCircuit.{conjunction, disjunction, neg}

/** Finite palindrome-prefix flag worker on an actual frozen mirror view.
  * The fixture loads a word, marks its lower endpoint with |, starts with !,
  * and executes local instructions with dots, building a persistent result stack.
  */
final class GSFlags(val circuit: Circuit, val prefix: String = "flags", val dual: Boolean = false) {
  val program: Program = if (dual) { GsDualFlags.compileDualFlags() } else { GsFlagHeads.compileFlags() }
  val readers: ReaderLiveness = GsHeadLiveness.analyzeReaders(program)
  val distances = new LiveDistances(circuit, program,
    if (dual) { GsDualFlags.DUAL_HEADS } else { GsFlagHeads.FLAG_HEADS }, prefix + ".distance")
  private val names = Vector.tabulate(readers.registers)(i => s"$prefix.r$i")
  private val rawNames = if (dual) { names } else { names.flatMap(name => Vector(name + ".forward", name + ".reverse")) }
  val streams = new StreamBank(circuit, Vector(prefix + ".begin", prefix + ".end") ++ rawNames)
  val begin: StreamHead = streams.heads(prefix + ".begin")
  val end: StreamHead = streams.heads(prefix + ".end")
  val registers: Vector[OrientedHead | MirrorHead] = names.map { name =>
    if (dual) { new OrientedHead(streams, name) } else { new MirrorHead(streams, name) }
  }
  val heads: Map[String, OrientedHead | MirrorHead] = readers.colors.map { case (head, color) => head -> registers(color) }
  val pcKey: String = prefix + ".pc"
  val phaseKey: String = prefix + ".phase"
  var pc: Value[Int] = circuit.get(PREVIOUS, pcKey, program.code.indices.toVector, program.start)
  var phase: Value[String] = circuit.get(PREVIOUS, phaseKey, Vector("idle", "broadcast", "maintenance", "run", "done"), "idle")
  val queueKey: String = prefix + ".queue"
  val remainingKey: String = prefix + ".remaining"
  private val queueNames = streams.queues.queues.keys.toVector
  var queue: Value[String] = circuit.get(PREVIOUS, queueKey, queueNames, queueNames.head)
  var remaining: Value[Int] = circuit.get(PREVIOUS, remainingKey, Vector(0, 1, 2, 3), 0)
  val returnKey: String = prefix + ".return"
  var returnPhase: Value[String] = circuit.get(PREVIOUS, returnKey, Vector("idle", "run", "done"), "idle")
  val postInputKey: String = prefix + ".post_input"
  var postInput: Value[String] = circuit.get(PREVIOUS, postInputKey, Vector("idle", "run", "done"), "idle")
  val cellKey: String = prefix + ".input_cell"
  var cell: Ref = circuit.getRef(PREVIOUS, cellKey)
  val initializedKey: String = prefix + ".initialized"
  var initialized: Expr = circuit.get(PREVIOUS, initializedKey, Vector(false, true), false).eqTo(true)
  val h = new Counter(distances.pool, prefix + ".h", Some("cells"))
  private val extraCounters: Vector[Counter] = if (dual) { Vector.empty } else {
    Vector("n", "b1", "bh1").map(name => new Counter(distances.pool, prefix + "." + name, Some("cells")))
  }
  val loadingCounters: Vector[Counter] = Vector(h, distances.length) ++ extraCounters
  val flagPool = new StackPool(circuit, Vector("cells" -> 1), Vector(false, true))
  val flags = new Stack(flagPool, prefix + ".output", Some("cells"))
  var emitted: Expr = FALSE

  private def read(head: String, enabled: Expr): Value[Any] = heads(head) match {
    case cursor: OrientedHead => cursor.read(enabled)
    case cursor: MirrorHead => cursor.read(enabled)
  }

  private def copy(target: Int, source: Int, enabled: Expr): Unit = {
    (registers(target), registers(source)) match {
      case (a: OrientedHead, b: OrientedHead) => a.copyFrom(b, enabled)
      case (a: MirrorHead, b: MirrorHead) => a.copyFrom(b, enabled)
      case _ => throw new IllegalStateException("mixed flag reader views")
    }
  }

  private def startHead(name: String, snapshot: StreamHead, reversed: Boolean, enabled: Expr): Unit = {
    heads(name) match {
      case cursor: OrientedHead => cursor.start(snapshot, reversed, enabled)
      case cursor: MirrorHead => cursor.start(begin, end, reversed, enabled)
    }
  }

  def tick(symbols: Option[Value[Any]] = None): Unit = {
    val char = symbols.getOrElse(circuit.input())
    val arrival = disjunction(char.eqTo('a'), char.eqTo('b'))
    val mark = char.eqTo('|')
    val start = char.eqTo('!')
    val dot = char.eqTo('.')
    val reset = char.eqTo('^')
    circuit.require(disjunction(Vector("idle", "run", "done").map(phase.eqTo)*), disjunction(arrival, mark))
    circuit.require(disjunction(phase.eqTo("idle"), phase.eqTo("done")), start)
    val oldPhase = phase
    val run = conjunction(dot, oldPhase.eqTo("run"))
    val broadcast = conjunction(dot, oldPhase.eqTo("broadcast"))
    val maintenance = conjunction(dot, oldPhase.eqTo("maintenance"))
    val events = program.code.indices.map(i => conjunction(run, pc.eqTo(i))).toVector
    val nextPc = mutable.LinkedHashMap.empty[Int, Expr]
    val copies = mutable.LinkedHashMap.empty[(Int, Int), Expr]
    val moves = mutable.LinkedHashMap.empty[(Int, Int), Expr]
    val requests = mutable.LinkedHashMap.empty[String, Expr]
    var halt = FALSE
    val flagEvents = VectorMap(false -> mutable.ArrayBuffer.empty[Expr], true -> mutable.ArrayBuffer.empty[Expr])
    def edge(target: Int, guard: Expr): Unit = { nextPc(target) = disjunction(nextPc.getOrElse(target, FALSE), guard) }
    def request(name: String, guard: Expr): Unit = { requests(name) = disjunction(requests.getOrElse(name, FALSE), guard) }
    for (((enabled, Row(event, targets)), state) <- events.zip(program.code).zipWithIndex) {
      def branch(decision: Expr): Unit = {
        edge(targets(0), conjunction(enabled, neg(decision)))
        edge(targets(1), conjunction(enabled, decision))
      }
      event match {
        case Event.Symbols(left, right) =>
          val a = read(left, enabled)
          val b = read(right, enabled)
          circuit.require(conjunction(neg(a.eqTo(None)), neg(b.eqTo(None))), enabled)
          branch(a.equal(b))
        case Event.Equal(left, right) => branch(distances.equal(left, right))
        case Event.Less(left, right) => branch(distances.less(left, right))
        case Event.Halt => halt = disjunction(halt, enabled)
        case other =>
          edge(targets(0), enabled)
          other match {
            case Event.Copy(target, source) =>
              if (readers.after(state)(target)) {
                val pair = (readers.colors(target), readers.colors(source))
                if (pair._1 != pair._2) { copies(pair) = disjunction(copies.getOrElse(pair, FALSE), enabled) }
              }
            case Event.Move(batch) =>
              val Event.Movement(head, direction) = batch.head
              if (readers.after(state)(head)) {
                val key = (readers.colors(head), direction)
                moves(key) = disjunction(moves.getOrElse(key, FALSE), enabled)
              }
            case Event.Flag(value) => flagEvents(value) += enabled
            case _ => throw new IllegalArgumentException("unsupported GS flag instruction")
          }
      }
    }
    distances.execute(events)
    for (((register, direction), enabled) <- moves) {
      val (name, needed) = registers(register) match {
        case cursor: OrientedHead => cursor.move(direction, enabled)
        case cursor: MirrorHead => cursor.move(direction, enabled)
      }
      request(name, needed)
    }
    for (((target, source), enabled) <- copies) { copy(target, source, enabled) }
    flags.clear(start)
    for ((value, conditions) <- flagEvents) {
      flags.push(data = Some(Value.constant(value)), enabled = disjunction(conditions.toSeq*), slot = Some(0))
    }
    emitted = disjunction(flagEvents.valuesIterator.flatMap(_.iterator).toSeq*)
    begin.queue.push(cell, broadcast)
    request(begin.queue.name, broadcast)
    for ((name, target) <- streams.queues.queues) { target.workUnit(conjunction(maintenance, queue.eqTo(name))) }
    end.followArrival(NEW, arrival)
    cell = Ref.select(arrival, NEW, cell)
    postInput = Value.select(arrival, Value(Vector("idle", "run", "done").map(name => name -> oldPhase.eqTo(name))), postInput)
    if (!dual) {
      val initial = neg(initialized)
      extraCounters(0).inc(initial, Some(0))
      extraCounters(0).inc(arrival, Some(1))
      extraCounters(0).inc(arrival, Some(2))
      extraCounters(1).inc(initial, Some(3))
      extraCounters(1).inc(arrival, Some(4))
      extraCounters(2).inc(initial, Some(5))
      extraCounters(2).inc(arrival, Some(6))
      extraCounters(2).inc(arrival, Some(7))
    }
    h.inc(arrival, Some(8))
    distances.loadOne(arrival)
    h.reset(mark)
    if (!dual) { extraCounters(2).copyFrom(extraCounters(1), mark) }
    initialized = TRUE
    val entry: VectorMap[(String, String), Option[(Counter, Int)]] = if (dual) {
      startHead("Origin", begin, false, start)
      startHead("TextOrigin", end, true, start)
      startHead("OriginalEnd", begin, true, start)
      VectorMap(("Lower", "Upper") -> Some((h, -1)),
        ("Origin", "OriginalEnd") -> Some((distances.length, -1)),
        ("Origin", "Upper") -> Some((distances.length, -1)),
        ("OriginalEnd", "Lower") -> Some((h, 1)),
        ("OriginalEnd", "TextOrigin") -> Some((distances.length, 1)),
        ("OriginalEnd", "Upper") -> None)
    } else {
      startHead("Origin", begin, false, start)
      startHead("OriginalEnd", end, true, start)
      VectorMap(("Origin", "OriginalEnd") -> Some((extraCounters(0), -1)),
        ("OriginalEnd", "Lower") -> Some((extraCounters(2), 1)),
        ("Origin", "Upper") -> Some((distances.length, -1)),
        ("OriginalEnd", "Upper") -> Some((extraCounters(1), 1)),
        ("Lower", "Upper") -> Some((h, -1)))
    }
    distances.initialize(entry, start)
    pc = Value.select(conjunction(run, neg(halt)), Value(nextPc), pc)
    pc = Value.select(start, Value.constant(program.start), pc)
    val finishedWork = conjunction(maintenance, remaining.eqTo(1))
    remaining = Value.select(maintenance, remaining.map(n => math.max(0, n - 1)), remaining)
    phase = Value.select(finishedWork, returnPhase, phase)
    val needed = disjunction(requests.values.toSeq*)
    queue = Value.select(needed, Value(requests), queue)
    remaining = Value.select(needed, Value.constant(3), remaining)
    returnPhase = Value.select(needed, Value.select(broadcast, postInput, Value.constant("run")), returnPhase)
    phase = Value.select(needed, Value.constant("maintenance"), phase)
    phase = Value.select(arrival, Value.constant("broadcast"), phase)
    phase = Value.select(start, Value.constant("run"), phase)
    phase = Value.select(halt, Value.constant("done"), phase)
    begin.reset(reset)
    end.reset(reset)
    loadingCounters.foreach(_.reset(reset))
    flags.clear(reset)
    phase = Value.select(reset, Value.constant("idle"), phase)
    initialized = conjunction(initialized, neg(reset))
  }

  def commit(): Unit = {
    distances.commit()
    loadingCounters.filterNot(_ eq distances.length).foreach(_.commit())
    streams.commit()
    registers.foreach {
      case cursor: OrientedHead => cursor.commit()
      case cursor: MirrorHead => cursor.commit()
    }
    flags.commit()
    for ((key, value) <- Vector[(String, Value[Any])](pcKey -> pc, phaseKey -> phase, queueKey -> queue,
      remainingKey -> remaining, returnKey -> returnPhase, postInputKey -> postInput)) { circuit.put(key, value) }
    circuit.put(initializedKey, Value.select(initialized, Value.constant(true), Value.constant(false)))
    circuit.putRef(cellKey, cell)
    circuit.put(prefix + ".emitted", Value.select(emitted, Value.constant(true), Value.constant(false)), Vector(false, true), false)
  }
}

object ScaffoldGsFlags {
  def fixture(dual: Boolean = false): (Circuit, GSFlags, Scaffold) = {
    val circuit = new Circuit("ab|!.^")
    circuit.put("input", Value.select(circuit.input().eqTo('b'), Value.constant('b'), Value.constant('a')), Vector('a', 'b'), 'a')
    val worker = new GSFlags(circuit, dual = dual)
    worker.tick()
    worker.commit()
    val (_, answer) = worker.flags.peek()
    (circuit, worker, circuit.machine(conjunction(worker.phase.eqTo("done"), neg(worker.flags.empty()), answer.eqTo(true))))
  }
}

File: docs/palindromes-in-peg/scaffold_gs_flags.py
"""Finite palindrome-prefix flag worker on an actual frozen mirror view.

The fixture loads a binary word, | marks the lower interval endpoint, !
starts the offline worker, and dots execute local instructions. Upper is
the loaded word's length. The worker builds a persistent result stack.
"""
from gs_flag_heads import compile_flags, FLAG_HEADS
from gs_dual_flags import compile_dual_flags, DUAL_HEADS
from gs_heads import TESTS
from gs_head_liveness import analyze_readers
from scaffold_circuit import (Circuit, Value, Ref, NEW, PREVIOUS, TRUE, FALSE,
                              conjunction, disjunction, neg, choose)
from scaffold_circuit_structs import Counter, Stack, StackPool
from scaffold_live_distances import LiveDistances
from scaffold_stream_heads import StreamBank, MirrorHead, OrientedHead


class GSFlags:
  def __init__(self, circuit, prefix="flags", dual=False):
    self.circuit, self.prefix = circuit, prefix
    self.dual = dual
    self.program = compile_dual_flags() if dual else compile_flags()
    self.readers = analyze_readers(self.program)
    self.distances = LiveDistances(circuit, self.program, DUAL_HEADS if dual else FLAG_HEADS, prefix + ".distance")
    names = tuple(f"{prefix}.r{i}" for i in range(self.readers.registers))
    raw_names = names if dual else tuple(name + suffix for name in names for suffix in (".forward", ".reverse"))
    self.streams = StreamBank(circuit, (prefix + ".begin", prefix + ".end", *raw_names))
    self.begin = self.streams.heads[prefix + ".begin"]
    self.end = self.streams.heads[prefix + ".end"]
    self.registers = [(OrientedHead if dual else MirrorHead)(self.streams, name) for name in names]
    self.heads = {head: self.registers[color] for head, color in self.readers.colors.items()}
    self.pc_key, self.phase_key = prefix + ".pc", prefix + ".phase"
    self.pc = circuit.get(PREVIOUS, self.pc_key, tuple(range(len(self.program.code))), self.program.start)
    self.phase = circuit.get(PREVIOUS, self.phase_key, ("idle", "broadcast", "maintenance", "run", "done"), "idle")
    self.queue_key, self.remaining_key = prefix + ".queue", prefix + ".remaining"
    qnames = tuple(self.streams.queues.queues)
    self.queue = circuit.get(PREVIOUS, self.queue_key, qnames, qnames[0])
    self.remaining = circuit.get(PREVIOUS, self.remaining_key, (0, 1, 2, 3), 0)
    self.return_key = prefix + ".return"
    self.return_phase = circuit.get(PREVIOUS, self.return_key, ("idle", "run", "done"), "idle")
    self.post_input_key = prefix + ".post_input"
    self.post_input = circuit.get(PREVIOUS, self.post_input_key, ("idle", "run", "done"), "idle")
    self.cell_key = prefix + ".input_cell"
    self.cell = circuit.get_ref(PREVIOUS, self.cell_key)
    self.initialized_key = prefix + ".initialized"
    self.initialized = circuit.get(PREVIOUS, self.initialized_key, (False, True), False).eq(True)
    self.h = Counter(self.distances.pool, prefix + ".h", "cells")
    self.loading_counters = [self.h, self.distances.length]
    if not dual:
      self.n = Counter(self.distances.pool, prefix + ".n", "cells")
      self.b1 = Counter(self.distances.pool, prefix + ".b1", "cells")
      self.bh1 = Counter(self.distances.pool, prefix + ".bh1", "cells")
      self.loading_counters.extend((self.n, self.b1, self.bh1))
    self.flag_pool = StackPool(circuit, {"cells": 1}, (False, True))
    self.flags = Stack(self.flag_pool, prefix + ".output", "cells")
    self.emitted = FALSE

  def tick(self, symbols=None):
    c = self.circuit
    char = c.input() if symbols is None else symbols
    arrival = disjunction(char.eq("a"), char.eq("b"))
    mark, start, dot, reset = char.eq("|"), char.eq("!"), char.eq("."), char.eq("^")
    c.require(disjunction(*(self.phase.eq(name) for name in ("idle", "run", "done"))), disjunction(arrival, mark))
    c.require(disjunction(self.phase.eq("idle"), self.phase.eq("done")), start)
    phase = self.phase
    run, broadcast, maintenance = (conjunction(dot, phase.eq(name)) for name in ("run", "broadcast", "maintenance"))
    events = [conjunction(run, self.pc.eq(i)) for i in range(len(self.program.code))]
    next_pc, copies, moves, requests = {}, {}, {}, {}
    halt, flags = FALSE, {False: [], True: []}

    def edge(target, guard):
      next_pc[target] = disjunction(next_pc.get(target, FALSE), guard)

    def request(queue, guard):
      requests[queue] = disjunction(requests.get(queue, FALSE), guard)

    for state, (enabled, (event, targets)) in enumerate(zip(events, self.program.code)):
      op, *args = event
      if op in TESTS:
        if op == "symbols":
          a, b = (self.heads[name].read(enabled) for name in args)
          c.require(conjunction(neg(a.eq(None)), neg(b.eq(None))), enabled)
          decision = a.equal(b)
        else:
          decision = (self.distances.equal if op == "equal" else self.distances.less)(*args)
        edge(targets[0], conjunction(enabled, neg(decision)))
        edge(targets[1], conjunction(enabled, decision))
      elif op == "halt":
        halt = disjunction(halt, enabled)
      else:
        edge(targets[0], enabled)
        if op == "copy":
          target, source = args
          if target in self.readers.after[state]:
            pair = self.readers.colors[target], self.readers.colors[source]
            if pair[0] != pair[1]:
              copies[pair] = disjunction(copies.get(pair, FALSE), enabled)
        elif op == "move":
          head, direction = args[0][0]
          if head in self.readers.after[state]:
            key = self.readers.colors[head], direction
            moves[key] = disjunction(moves.get(key, FALSE), enabled)
        elif op == "flag":
          flags[args[0]].append(enabled)
        else:
          raise ValueError("unsupported GS flag instruction")
    self.distances.execute(events)
    for (register, direction), enabled in moves.items():
      queue, needed = self.registers[register].move(direction, enabled)
      request(queue, needed)
    for (target, source), enabled in copies.items():
      self.registers[target].copy_from(self.registers[source], enabled)
    self.flags.clear(start)
    for value, conditions in flags.items():
      self.flags.push(data=Value.constant(value), enabled=disjunction(*conditions), slot=0)
    self.emitted = disjunction(*(condition for conditions in flags.values() for condition in conditions))

    self.begin.queue.push(self.cell, broadcast)
    request(self.begin.queue.name, broadcast)
    for name, queue in self.streams.queues.queues.items():
      queue.work_unit(conjunction(maintenance, self.queue.eq(name)))
    self.end.follow_arrival(NEW, arrival)
    self.cell = Ref.select(arrival, NEW, self.cell)
    self.post_input = Value.select(arrival, Value({name: phase.eq(name) for name in ("idle", "run", "done")}), self.post_input)
    # Entry distances are built by unit updates while loading; no length
    # or interval integer is injected when starting a job.
    if not self.dual:
      initial = neg(self.initialized)
      self.n.inc(initial, slot=0)
      self.n.inc(arrival, slot=1)
      self.n.inc(arrival, slot=2)
      self.b1.inc(initial, slot=3)
      self.b1.inc(arrival, slot=4)
      self.bh1.inc(initial, slot=5)
      self.bh1.inc(arrival, slot=6)
      self.bh1.inc(arrival, slot=7)
    self.h.inc(arrival, slot=8)
    self.distances.load_one(arrival)
    self.h.reset(mark)
    if not self.dual:
      self.bh1.copy_from(self.b1, mark)
    self.initialized = TRUE

    if self.dual:
      self.heads["Origin"].start(self.begin, False, start)
      self.heads["TextOrigin"].start(self.end, True, start)
      self.heads["OriginalEnd"].start(self.begin, True, start)
      entry = {("Lower", "Upper"): (self.h, -1),
               ("Origin", "OriginalEnd"): (self.distances.length, -1),
               ("Origin", "Upper"): (self.distances.length, -1),
               ("OriginalEnd", "Lower"): (self.h, 1),
               ("OriginalEnd", "TextOrigin"): (self.distances.length, 1),
               ("OriginalEnd", "Upper"): None}
    else:
      self.heads["Origin"].start(self.begin, self.end, False, start)
      self.heads["OriginalEnd"].start(self.begin, self.end, True, start)
      entry = {("Origin", "OriginalEnd"): (self.n, -1),
                               ("OriginalEnd", "Lower"): (self.bh1, 1),
                               ("Origin", "Upper"): (self.distances.length, -1),
                               ("OriginalEnd", "Upper"): (self.b1, 1),
                               ("Lower", "Upper"): (self.h, -1)}
    self.distances.initialize(entry, start)
    self.pc = Value.select(conjunction(run, neg(halt)), Value(next_pc), self.pc)
    self.pc = Value.select(start, Value.constant(self.program.start), self.pc)
    finished_work = conjunction(maintenance, self.remaining.eq(1))
    self.remaining = Value.select(maintenance, self.remaining.map(lambda n: max(0, n - 1)), self.remaining)
    self.phase = Value.select(finished_work, self.return_phase, self.phase)
    needed = disjunction(*requests.values())
    self.queue = Value.select(needed, Value(requests), self.queue)
    self.remaining = Value.select(needed, Value.constant(3), self.remaining)
    self.return_phase = Value.select(needed, Value.select(broadcast, self.post_input, Value.constant("run")), self.return_phase)
    self.phase = Value.select(needed, Value.constant("maintenance"), self.phase)
    self.phase = Value.select(arrival, Value.constant("broadcast"), self.phase)
    self.phase = Value.select(start, Value.constant("run"), self.phase)
    self.phase = Value.select(halt, Value.constant("done"), self.phase)
    self.begin.reset(reset)
    self.end.reset(reset)
    for counter in self.loading_counters:
      counter.reset(reset)
    self.flags.clear(reset)
    self.phase = Value.select(reset, Value.constant("idle"), self.phase)
    self.initialized = conjunction(self.initialized, neg(reset))

  def finalize(self):
    self.distances.finalize()
    for counter in self.loading_counters:
      if counter is not self.distances.length:
        counter.finalize()
    self.streams.finalize()
    for head in self.registers:
      head.finalize()
    self.flags.finalize()
    for key, value in ((self.pc_key, self.pc), (self.phase_key, self.phase),
                       (self.queue_key, self.queue), (self.remaining_key, self.remaining),
                       (self.return_key, self.return_phase), (self.post_input_key, self.post_input)):
      self.circuit.put(key, value)
    self.circuit.put(self.initialized_key, Value.select(self.initialized, Value.constant(True), Value.constant(False)))
    self.circuit.put_ref(self.cell_key, self.cell)
    self.circuit.put(self.prefix + ".emitted", Value.select(self.emitted, Value.constant(True), Value.constant(False)),
                     (False, True), False)


def fixture(dual=False):
  circuit = Circuit("ab|!.^")
  circuit.put("input", Value.select(circuit.input().eq("b"), Value.constant("b"), Value.constant("a")), tuple("ab"), "a")
  worker = GSFlags(circuit, dual=dual)
  worker.tick()
  worker.finalize()
  _, answer = worker.flags.peek()
  machine = circuit.machine(conjunction(worker.phase.eq("done"), neg(worker.flags.empty()), answer.eq(True)))
  return circuit, worker, machine
